src/puya/ir/builder/main.py

```diff
@@ -504,11 +520,17 @@ def visit_method_constant(self, expr: puya.awst.nodes.MethodConstant) -> TExpres
         return MethodConstant(value=expr.value, source_location=expr.source_location)
 
     def visit_tuple_expression(self, expr: awst_nodes.TupleExpression) -> TExpression:
-        items = []
+        items = list[Value]()
         for item in expr.items:
-            # TODO: don't rely on a pure function's side effects (raising) for validation
-            wtype_to_ir_type(item)
-            items.append(self.visit_and_materialise_single(item))
+            if isinstance(item.wtype, wtypes.WTuple):
+                nested = item.accept(self)
+                assert isinstance(nested, ValueTuple)
```

Comment:
how are you planning to handle other multi-valued types? such as user subroutines or intrinsics that return multiple values?